File: CS174/Asn2/hangman.py


words = ['cow', 'horse', 'deer', 'elephant', 'lion', 'tiger', 'baboon', 'donkey', 'fox', 'giraffe']
MAX_ERRORS = 6

class Hangman(object):
		def __init__(self):
			self.parts = [' ------------', ' |         |', ' |          O', ' |         / |', ' |          |', ' |         / |']
			self.errors = 0
		
		def update(self):
			self.errors += 1
			for part in self.parts[:self.errors]:
				print(part)

class Word(object):
		def __init__(self, word):
			self.name = word
			self.length = len(word)
			self.output = ['_'] * self.length
		
		def takeGuess(self, guess):
			if guess in self.name:
				self.update(guess)
				print("The letter is in the word.")
				return True
			else:
				print("The letter is not in the word.")
				return False
		
		def update(self, guess):
			for ii in range(self.length):
				if self.name[ii] == guess:
					self.output[ii] = guess
		
		def show(self):
			print("Letters matched so far:", ''.join(self.output))
		
		def isFound(self):
			if '_' in self.output: return False
			return True

def checkValidity(ans):
		if not ans:
			print("Empty input!")
			return False
		try: 
			ans = int(ans)
		except ValueError:
			print("Input must be an integer!")
			return False
		if not 0 <= ans < 10:
			print("Index is out of range!")
			return False
		return ans

def intro():
		print("Welcome to Hangman! Guess the mystery word with less than", str(MAX_ERRORS), "mistakes!")
		while True:
			ans = input("Please enter an integer number (0<=number<" + str(len(words)) + ") to choose the word in the list: ")
			wordIndex = checkValidity(ans)
			if not wordIndex: continue
			word = Word(words[wordIndex])
			print("The length of the word is:", str(word.length))
			return word

def play(word):
		man = Hangman()
		while True:
			guess = input("\nPlease enter the letter you guess: ")
			if not guess.isalpha() or len(guess) != 1:
				print("You need to input a single alphabetic character!")
				continue
			guess = guess.lower()
			isMatch = word.takeGuess(guess)
			word.show()
			if word.isFound():
				print("You have found the mystery word. You win!")
				return
			elif not isMatch:
				man.update()
				if man.errors == MAX_ERRORS:
					print(' |\n |')
					print("Too many incorrect guesses. You lost!")
					print("The word was:", word.name + '.')
					return


word = intro()
play(word)
print("Goodbye!")

File: CS174/Asn2/hangman.clj

; messages
(def intro "Welcome to Hangman! Guess the mystery word with less than 6 mistakes!")
(def options "Please enter an integer number (0<=number<10) to choose the word in the list: ")
(def hint "The length of the word is: ")
(def instruct "Please enter the letter you guess: ")
(def hit "The letter is in the word.")
(def miss "The letter is not in the word.")
(def progress "Letters matched so far: ")
(def bye "Goodbye!")

; error messages
(def empty-input "Empty input!")
(def coerce-error "Input must be an integer!")
(def range-error "Index is out of range!")
(def extra-chars "You need to input a single alphabetic character!")

; target words
(def words ["cow", "horse", "deer", "elephant", "lion", "tiger", "baboon", "donkey", "fox", "giraffe"])
(def max_value 9)
(def min_value 0)


(defn input 
	([] 
		(println "Please enter your input: ")
		(read-line)
	)
	([user-message]
		(println user-message)
		(read-line)
	)
)

(defn coerce| 
	([value]
		(try
			(Integer/parseInt value)
			(catch Exception e false)
		)
	)
	([value verbose]
		(let [attempt (coerce| value)]	
			(if
				(boolean attempt)
				attempt
				(do
					(println coerce-error)
					false
				)
			)
		)
	)
)

(defn between? 
	([x] (and (> x min_value) (< x max_value)))
	([x verbose]
		(if 
			(between? x)
			true
			(do
				(println range-error)
				false
			)
		)
	)
)
	

(defn passes? [x]
	(let [attempt (coerce| x "verbose")] 
		(cond
			(or  
				(not attempt)
				(not (between? attempt "verbose"))
			) false
			:else attempt
		)
	)	
)

(defn menu []
	(loop [choice (input options)]
		(let [attempt (passes? choice)]
			(if attempt
				attempt
				(recur (input options))
			)
		)
	)
)

(defn play-game [option]
	(seq (println "playing game"))
)


; main function
(do
	(println intro)
	(play-game (menu))
	(println bye)
)
